### PR TITLE
Package resources + a metadata file into a single .zip file.

### PR DESCRIPTION
**Pull request by JF002**
Files (fonts and pictures) are converted and packed into a single .zip file using a CMake `custom_target`. This target

- Converts the fonts into .bin files (using a slightly modified version of `src/displayapp/fonts/generate.py`)
- Converts the images into .bin files (using a new script inspired from `generate.py` above)
- Generates a .json file with metadata (see below)
- Zip everything in a single `infinitime-resources-1.11.0.zip`

The content of the zip file looks like this:

```
infinitime-resources-1.11.0.zip
|
|---- 7segments_40.bin
|---- 7segments_115.bin
|---- infineat-1bin
...
|---- resources.json
```

resources.json:

```
{
    "resources": [
        {
            "filename": "7segments_40.bin",
            "path": "/fonts/7segments_40.bin"
        },
        {
            "filename": "7segments_115.bin",
            "path": "/fonts/7segments_115.bin"
        },i
        {
            "filename": "infineat-1.bin",
            "path": "/fonts/infineat-1.bin"
        }
    ],
    "obsolete_files": [
        {
            "path": "/background_v1.bin",
            "since": "1.11.0"
        },
        {
            "path": "/font_v2.bin",
            "since": "1.11.0"
        }
    ]
}
```

- The array `resources` contains the list of files from the .zip file that must be uploaded to the watch. It contains the name of the file in the .zip file and the path (filename) where it must be written to the watch.
- The array `obsolete_files` contains a list of files that can optionally remove from the watch. This is a hint to the companion app which lists the files that are not used anymore in this version of infinitime and that can be removed. It's up to the companion app to remove them to free some memory space. Each item contain the filename in the watch FS and the version of infinitime that made that file obsolete.

EDIT : here is an example package: [infinitime-resources-1.10.0.zip](https://github.com/InfiniTimeOrg/InfiniTime/files/9500512/infinitime-resources-1.10.0.zip)

And this is the expected result once the files are uploaded (screenshots using [ITD](https://gitea.arsenm.dev/Arsen6331/itd):
image

```
./itctl fs ls                            
d 0.0 B  .
d 0.0 B  ..
- 4.9 kB 7segments_115.bin
- 760 B  7segments_40.bin
- 4.4 kB bebas.bin
- 1.4 kB infineat-1.bin
- 1.8 kB lv_font_dots_40.bin
- 115 kB matrix.bin
-  40 B  settings.dat
- 440 B  teko.bin

```

To support this, companion apps will have to implement our [BLE FS API](https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/BLEFS.md).

See [this comment](https://github.com/InfiniTimeOrg/InfiniTime/issues/1285#issuecomment-1221605579)

### CODE REVIEW

File: src/resources/generate-fonts.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self, d):
+        self.file = d['file']
+        if not os.path.exists(self.file):
+            self.file = os.path.join(os.path.dirname(sys.argv[0]), self.file)
+        self.range = d.get('range')
+        self.symbols = d.get('symbols')
+
+
+def gen_lvconv_line(lv_font_conv: str, dest: str, size: int, bpp: int, format: str, sources: typing.List[Source], compress:bool=False):
+    if format != "lvgl" and format != "bin":
```
Comment (yehoshuapw) [inline src/resources/generate-fonts.py:22]:
```suggestion
    if format != "lvgl" and format != "bin":
        format = "bin" if dest.lower().endswith(".bin") else "lvgl"
```

File: src/resources/generate-fonts.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        format = "lvgl"
```
Comment (yehoshuapw) [inline src/resources/generate-fonts.py:23]:
```suggestion
    if format != "lvgl" and format != "bin":
        format = "bin" if dest.lower().endswith(".bin") else "lvgl"
```

**Comment by Arsen6331** [comment]
I'm currently implementing this for ITD. Should be ready soon if everything goes well. I'll make a branch with it when I'm done and then merge when this is released.

**Comment by Arsen6331** [comment]
One question that I have is: should I load the resources (delete obsolete, upload new, etc.) before or after the DFU itself?

**Comment by JF002** [comment]
> should I load the resources (delete obsolete, upload new, etc.) before or after the DFU itself?

That's a good question! Loading the new resources (and potentially removing older ones) before the firmware update might break the firmware that is currently running, if it's still using the resources that will be deleted. Loading them after the firmware update might break the new firmware if it tries to load resources that do not exist yet... 
This is actually one of to TODO task that need to be done before releasing the functionality : "*InfiniTime must handle the case where an app/watchface needs a resource from the FS and the corresponding file does not exist*" https://github.com/InfiniTimeOrg/InfiniTime/issues/321#issuecomment-1177289430

In the end, I think we need to ensure that the firmware (app & watchface) will never crash or misbehave if the resource does not exist. In this case, the update sequence (firmware then resource or resource then firmware) will be the choice of the user or the companion app.

**Comment by JF002** [comment]
Here is a new resource package that maps files into subdirectories (`/fonts` and `/images`). It might be useful to ensure that implementations in companion apps handle the subdirectories correctly

[infinitime-resources-1.10.0.zip](https://github.com/InfiniTimeOrg/InfiniTime/files/9500512/infinitime-resources-1.10.0.zip)

File: src/resources/generate-fonts.py (hunk 2 above)
Comment (JF002) [reply src/resources/generate-fonts.py]: Done, thanks!

**Comment by JF002** [comment]
So my plan for this feature is:
  - Merge this branch to allow generation of resources
  - Extend Infineat to support external resource and merge it. It also brings Infineat watchface and the possibility to select more than 4 watchfaces in the settings.
  - Extend and merge the G7710 watchface
  - Talk with companion app developers and contributors hoping they'll add support for this feature in companion apps. ITD already supports it, and I'm currently working on the integration in Amazfish.
  - If anything goes wrong, we'll be able to temporarily disable the new watchfaces that need additional resources in the settings.
  - Release and enjoy!

**Pull request merged**